saleor/checkout/problems.py

```diff
@@ -0,0 +1,185 @@
+import datetime
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Iterable, Optional, Tuple, Union
+
+import pytz
+
+from ..graphql.channel import ChannelContext
+from ..product.models import ProductChannelListing, ProductVariant
+from ..warehouse.models import Stock
+from .fetch import CheckoutInfo, CheckoutLineInfo
+from .models import CheckoutLine
+
+
+@dataclass
+class CheckoutLineProblemInsufficientStock:
+    available_quantity: int
+    line: CheckoutLine
+    variant: Optional[ChannelContext[ProductVariant]] = None
+
+
+@dataclass
+class CheckoutLineProblemVariantNotAvailable:
+    line: CheckoutLine
+
+
+CHECKOUT_LINE_PROBLEM_TYPE = Union[
+    CheckoutLineProblemInsufficientStock, CheckoutLineProblemVariantNotAvailable
+]
+CHECKOUT_PROBLEM_TYPE = Union[
+    CheckoutLineProblemInsufficientStock, CheckoutLineProblemVariantNotAvailable
+]
+
+VARIANT_ID = int
+PRODUCT_ID = int
+CHANNEL_SLUG = str
+COUNTRY_CODE = str
+
+CHECKOUT_LINE_PK_TYPE = str
+
+
+def get_insufficient_stock_lines(
+    lines: Iterable["CheckoutLineInfo"],
+    variant_stock_map: dict[
+        tuple[
+            VARIANT_ID,
+            CHANNEL_SLUG,
+            COUNTRY_CODE,
+        ],
+        Iterable[Stock],
+    ],
+    country_code: str,
+) -> list[Tuple["CheckoutLineInfo", int]]:
+    """Return checkout lines with insufficient stock."""
+    variant_to_quantity_map: dict[int, int] = defaultdict(int)
+    variant_to_available_quantity_map: dict[int, int] = defaultdict(int)
+    for line_info in lines:
+        variant_to_quantity_map[line_info.variant.id] += line_info.line.quantity
+        variant_stocks = variant_stock_map.get(
+            (line_info.variant.id, line_info.channel.slug, country_code), []
+        )
+        variant_to_available_quantity_map[line_info.variant.id] = sum(
+            [
+                stock.available_quantity  # type: ignore[attr-defined]
+                for stock in variant_stocks
+            ]
+        )
+    insufficient_stocks = []
+    for line_info in lines:
+        if not line_info.variant.track_inventory:
+            continue
+        quantity = variant_to_quantity_map[line_info.variant.id]
+        available_quantity = variant_to_available_quantity_map[line_info.variant.id]
+        if available_quantity < quantity:
+            insufficient_stocks.append((line_info, available_quantity))
+    return insufficient_stocks
+
+
+def get_not_available_lines(
+    lines: Iterable["CheckoutLineInfo"],
+    product_channel_listings_map: dict[
+        tuple[
+            PRODUCT_ID,
+            CHANNEL_SLUG,
+        ],
+        ProductChannelListing,
+    ],
+):
+    lines_not_available = []
+    now = datetime.datetime.now(pytz.UTC)
+    for line in lines:
```

Comment:
Did you consider extracting those ifs to a method to avoid that many `continue` statements?

Reply:
Extracted 👍